torchsde/_core/adjoint_sde.py

```diff
@@ -58,32 +58,72 @@ def __init__(self,
             NOISE_TYPES.diagonal: self.gdg_prod_diagonal,
         }.get(sde.noise_type, self.gdg_prod_default)
 
-    def _unpack_y_aug(self, y_aug):
+    def _unpack_y_aug(self, t, y_aug, v):
+        # These leaf checks are very important.
+        # _unpack_y_aug is used where we want to compute:
+        # ```
+        # with torch.enable_grad():
+        #     s = some_function(y)
+        #     torch.autograd.grad(s, [y] + params, ...)
+        # ```
+        # where `some_function` implicitly depends on `params`.
+        # However if y has history of its own then in principle it could _also_ depend upon `params`, and this call to
+        # `grad` will go all the way back to that.
+        # To avoid this, we require that y (and the other tensors) be leaf tensors.
+        #
+        # Note that this is also the reason for the `y0.detach()` in adjoint.py::_SdeintAdjointMethod.forward. Unless
+        # we detach then y0 may have a history and these checks will fail. Note that this isn't then producing the wrong
+        # gradients when backprop'ing as usual because `torch.autograd.Function.forward` has an implicit
+        # `torch.no_grad()` guard, i.e. any history doesn't affect the usual gradient computations anyway. But it does
+        # still get detected if we call `torch.autograd.grad` ourselves, which is explicitly what we don't want as per
+        # the previous paragraph.
+        assert t.is_leaf, "Internal error: please report a bug to torchsde"
+        assert y_aug.is_leaf, "Internal error: please report a bug to torchsde"
+        assert v.is_leaf, "Internal error: please report a bug to torchsde"
```

Comment:
I think if we assume `v` comes from the Brownian motion, then it's also safe to assume that it's a leaf, since tensors generated by BM doesn't require gradients, and so is the returning tensor of any function of such tensors.

Reply:
Regarding this and the points about `t`: AFAIK both `t` and `v` should never require gradients in the current set-up.

But that's a detail that's beyond `AdjointSDE`'s scope. `AdjointSDE` requires that these conditions are satisfied, so it should enforce them. Practically: if that detail ever changes down the line then I'd prefer to know about it without things silently going wrong.